rules/providers.bzl

```diff
@@ -0,0 +1,109 @@
+ScalaConfiguration = provider(
+    doc = "Provides access to the Scala compiler jars",
+    fields = {
+        "version": "the Scala full version",
+        "compiler_classpath": "the compiler classpath",
+        "runtime_classpath": "the minimal runtime classpath",
+    },
+)
+
+def _declare_scala_configuration_implementation(ctx):
+    return [ScalaConfiguration(
+        version = ctx.attr.version,
+        compiler_classpath = ctx.files.compiler_classpath,
+        runtime_classpath = ctx.attr.runtime_classpath,
+    )]
+
+declare_scala_configuration = rule(
+    implementation = _declare_scala_configuration_implementation,
+    attrs = {
+        "version": attr.string(
+            mandatory = True,
+        ),
+        "compiler_classpath": attr.label_list(
+            mandatory = True,
+            providers = [JavaInfo],
+        ),
+        "runtime_classpath": attr.label_list(
+            mandatory = True,
+            providers = [JavaInfo],
+        ),
+    },
+)
+
+ScalaInfo = provider(
+    doc = "Provider for cross versioned scala rule outputs",
+    fields = {
+        "scala_configuration": "the scala configuration associated with this output",
+    },
+)
+
+ZincConfiguration = provider(
+    doc = "Provides additional items needed by Zinc",
+    fields = {
+        "compiler_bridge": "the compiled Zinc compiler bridge",
+    },
+)
+
+def _declare_zinc_configuration_implementation(ctx):
+    return [ZincConfiguration(
+        compiler_bridge = ctx.files.compiler_bridge,
+    )]
+
+declare_zinc_configuration = rule(
+    implementation = _declare_zinc_configuration_implementation,
+    attrs = {
+        "compiler_bridge": attr.label(
+            allow_single_file = True,
+            mandatory = True,
+        ),
+    },
+)
+
+ZincInfo = provider(
+    doc = "Provides additional outputs from Zinc",
+    fields = {
+        "analysis": "Zinc analysis file in binary form",
+    },
+)
+
+def _collect(index, entries):
+    return [
+        entry[index]
+        for entry in entries
+        if index in entry
+    ]
+
+def _join_configurations_implementation(ctx):
+    return (
+        _collect(ScalaConfiguration, ctx.attr.configurations) +
+        _collect(ZincConfiguration, ctx.attr.configurations)
+    )
+
+join_configurations = rule(
+    implementation = _join_configurations_implementation,
+    attrs = {
+        "configurations": attr.label_list(
+            mandatory = True,
+            providers = [[ScalaConfiguration], [ZincConfiguration]],
+        ),
+    },
+)
+
+# TODO: move these to another file?
+# TODO: implement these with an aspect?
```

Comment:
#51 implements JarsToLabels with an aspect.

IntelliJInfo is more complicated.
They mostly use old-style structs rather than the new providers.
I have a hack in place because I think they should be querying for the transitive deps, rather than just the direct. So I write the transitive deps to old-style struct java.transitive_exports which they look at.

Reply:
👍